Add dynamic shape support to CoreML

metascroy opened the pull request:
Summary:
Dynamic shape currently lead to runtime errors in CoreML.

The AOT flow prepares the CoreML pte to work with dynamic shapes.  These inputs are correctly evaluated by the CoreML model, but a runtime error occurs when copying the model outputs back to ET.

This is because the model outputs have shape that is smaller than the outputArg MultiArray and the copy fails.

This PR resizes the output args to account for dynamic shapes in the ET CoreML delegate.

Note that outputBackings appear to fail in the dynamic case, likely because the outputBacking tensor is bigger than it should be.  Ideally CoreML would not fail as long as the outputBacking had enough storage to hold the output since we don't know the expected size until after the model runs.  The output is resized after the model is run anyway.

In any case, dynamic shapes take the "ignoreOutputBackings" route.

cc @kimishpatel @YifanShenSZ @cymbalrush

metascroy committed:
Add dynamic shape support to CoreML

Summary:
Dynamic shape currently lead to runtime errors in CoreML.

The AOT flow prepares the CoreML pte to work with dynamic shapes.  These inputs are correctly evaluated by the CoreML model, but a runtime error occurs when copying the model outputs back to ET.

This is because the model outputs have shape that is smaller than the outputArg MultiArray and the copy fails.

This PR resizes the output args to account for dynamic shapes in the ET CoreML delegate.

Note that outputBackings appear to fail in the dynamic case, likely because the outputBacking tensor is bigger than it should be.  Ideally CoreML would not fail as long as the outputBacking had enough storage to hold the output since we don't know the expected size until after the model runs.  The output is resized after the model is run anyway.

In any case, dynamic shapes take the "ignoreOutputBackings" route.

cymbalrush reviewed (approved):
Looks good thank you!

cccclai commented:
Thanks - lint error needs to be fixed and it will be good to have a test coverage.

metascroy commented on CMakeLists.txt:
```diff
@@ -903,6 +903,10 @@ if(EXECUTORCH_BUILD_EXECUTOR_RUNNER)
   endif()
 
+  if(EXECUTORCH_BUILD_COREML)
+    list(APPEND _executor_runner_libs coremldelegate)
```
I can remove, it was just very convenient to have a C++ runner that works with CoreML pte files for debugging the existing crash.

cccclai replied:
Yeah I guess it's fine

metascroy commented:
> Thanks - lint error needs to be fixed and it will be good to have a test coverage.

Will add

metascroy committed:
up

metascroy committed:
up

metascroy commented:
> Thanks - lint error needs to be fixed and it will be good to have a test coverage.

Added new tests for resize.  In doing so, I noticed one of the existing tests is broken on main (diabled it in this PR).  We don't run these tests in CI, but I added an issue to enable them (https://github.com/pytorch/executorch/issues/9115).

metascroy commented:
@shoumikhin can you take a look at the error in building the benchmark app in the CI here.  It's not clear to me what the issue is.  Note that the frameworks build correctly.

metascroy committed:
Merge branch 'main' into coreml-dynamic-shape

metascroy committed:
Merge branch 'main' into coreml-dynamic-shape

metascroy commented:
@pytorchbot label "topic: not user facing"
